### PR TITLE
Generate serialization for WebCore::Color

### PR DESCRIPTION
**Pull request by sheeparegreat**
#### 71513d364e3d4cb1f32a6f8fcf8b387ce57e8ac2
<pre>
Generate serialization for WebCore::Color
<a href="https://bugs.webkit.org/show_bug.cgi?id=263089">https://bugs.webkit.org/show_bug.cgi?id=263089</a>
rdar://116882741

Reviewed by Ryosuke Niwa and Alex Christensen.

* Source/WebCore/platform/graphics/Color.cpp:
(WebCore::Color::Color):
(WebCore::Color::data const):
* Source/WebCore/platform/graphics/Color.h:
(WebCore::Color::encode const): Deleted.
(WebCore::Color::decode): Deleted.
* Source/WebCore/platform/graphics/ColorSpace.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/269336@main">https://commits.webkit.org/269336@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9fe166e320fbec01be3dae009b69ee933fa90ca2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22248 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23320 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24145 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20587 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22495 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26709 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22774 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21600 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19304 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24999 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19240 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20160 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26404 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20257 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20390 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24268 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20821 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17714 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20177 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24387 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2782 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20768 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/ColorTypes.h [hunk 1]
```diff
@@ -721,6 +736,26 @@ struct RGBA {
...
+    {
+        return (value >> 8) & 0xFF;
+    }
+
+    uint8_t alpha() const
+    {
+        return value & 0xFF;
+    }
+
     uint32_t value;
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/ColorTypes.h]: Let's just serialize this uint32_t instead of breaking it into pieces.  Then this will need no modification.

File: Source/WebCore/platform/graphics/ColorTypes.h [hunk 2]
```diff
@@ -201,6 +201,16 @@ constexpr bool operator==(const ColorType& a, const ColorType& b)
 
 // MARK: - RGB Color Types.
 
+enum class ColorSpace : uint8_t;
+
+struct OutOfLineColorData {
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/ColorTypes.h]: This should be a nested struct inside ColorData.

File: Source/WebCore/platform/graphics/Color.cpp [hunk 3]
```diff
@@ -49,6 +49,51 @@ Color::Color(Color&& other)
...
+
+        OutOfLineColorData oolcd = {
+            .colorSpace = colorSpace(),
+            .c1 = c1,
+            .c2 = c2,
+            .alpha = alpha
+        };
+
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
+                 .usesFunctionSerialization = flags().contains(FlagsIncludingPrivate::UseColorFunctionSerialization), // NOLINT
```
Comment (rniwa) [inline Source/WebCore/platform/graphics/Color.cpp]: Don't align lines like this. Always indent with 4 spaces instead.

File: Source/WebCore/platform/graphics/Color.cpp [hunk 4]
```diff
@@ -49,6 +49,51 @@ Color::Color(Color&& other)
...
+                 .data = oolcd }; // NOLINT
+
+    } else {
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
+                 .usesFunctionSerialization = flags().contains(FlagsIncludingPrivate::UseColorFunctionSerialization), // NOLINT
```
Comment (rniwa) [inline Source/WebCore/platform/graphics/Color.cpp]: Ditto.

File: Source/WebCore/platform/graphics/Color.cpp (hunk 3 above)
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.cpp]: Don't use NOLINT here.  Just make the indentation more like this:
return {
    something,
    somethingElse
};

File: Source/WebCore/platform/graphics/Color.h [hunk 5]
```diff
@@ -61,10 +61,17 @@ class Color {
         UseColorFunctionSerialization   = 1 << 1,
     };
 
+    struct ColorData {
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.h]: ColorDataForIPC?

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/graphics/ColorTypes.h [hunk 6]
```diff
@@ -201,6 +201,16 @@ constexpr bool operator==(const ColorType& a, const ColorType& b)
...
+    ColorSpace colorSpace;
+    float c1;
```
Comment (rniwa) [inline Source/WebCore/platform/graphics/ColorTypes.h]: Please spell out red, green, blue instead of c1, c2, c3.

**Review by rniwa — APPROVED** [review]

File: Source/WebCore/platform/graphics/ColorTypes.h (hunk 6 above)
Comment (weinig) [reply Source/WebCore/platform/graphics/ColorTypes.h]: This is not rgb data, it is just generic 3 channel color, so c1, c2, c3 is the canonical naming.

File: Source/WebCore/platform/graphics/ColorTypes.h [hunk 7]
```diff
@@ -711,6 +721,11 @@ struct RGBA {
     {
     }
 
+    explicit RGBA(uint8_t red, uint8_t green, uint8_t blue, uint8_t alpha)
```
Comment (weinig) [inline Source/WebCore/platform/graphics/ColorTypes.h]: Why is this needed?

Comment (achristensen07) [reply Source/WebCore/platform/graphics/ColorTypes.h]: It's not if we serialize the uint32_t instead.

File: Source/WebCore/platform/graphics/ColorTypes.h (hunk 2 above)
Comment (weinig) [inline Source/WebCore/platform/graphics/ColorTypes.h]: OutOfLineColorData is not part of the RGB color types family and should not be here. Please keep this isolated to WebCore::Color.

File: Source/WebCore/platform/graphics/Color.cpp [hunk 8]
```diff
@@ -49,6 +49,51 @@ Color::Color(Color&& other)
...
+Color::ColorData Color::data() const
+{
+    if (isOutOfLine()) {
+        auto& outOfLineComponents = asOutOfLine();
+        auto [c1, c2, c3, alpha] = outOfLineComponents.unresolvedComponents();
+
+        OutOfLineColorData oolcd = {
+            .colorSpace = colorSpace(),
+            .c1 = c1,
+            .c2 = c2,
```
Comment (sheeparegreat) [inline Source/WebCore/platform/graphics/Color.cpp:82]: .c3 = c3.

File: Source/WebCore/platform/graphics/Color.cpp [hunk 9]
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
...
 
+Color::Color(Color::ColorDataForIPC&& colorData)
+{
+    OptionSet<FlagsIncludingPrivate> flags;
+    if (colorData.isSemantic)
+        flags.add(FlagsIncludingPrivate::Semantic);
+    if (colorData.usesFunctionSerialization)
+        flags.add(FlagsIncludingPrivate::UseColorFunctionSerialization);
+
+    if (std::holds_alternative<PackedColor::RGBA>(colorData.data)) {
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.cpp]: This should use WTF::switchOn

File: Source/WebCore/platform/graphics/Color.cpp [hunk 10]
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
...
+        flags.add({ FlagsIncludingPrivate::Valid });
+
+        auto& d = std::get<PackedColor::RGBA>(colorData.data);
+        setColor(asSRGBA(d), flags);
+    } else if (std::holds_alternative<OutOfLineColorData>(colorData.data)) {
+        flags.add({ FlagsIncludingPrivate::Valid, FlagsIncludingPrivate::OutOfLine });
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.cpp]: This wasn't in Color::decode

File: Source/WebCore/platform/graphics/Color.cpp [hunk 11]
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
...
+Color::Color(Color::ColorDataForIPC&& colorData)
+{
+    OptionSet<FlagsIncludingPrivate> flags;
+    if (colorData.isSemantic)
+        flags.add(FlagsIncludingPrivate::Semantic);
+    if (colorData.usesFunctionSerialization)
+        flags.add(FlagsIncludingPrivate::UseColorFunctionSerialization);
+
+    if (std::holds_alternative<PackedColor::RGBA>(colorData.data)) {
+        flags.add({ FlagsIncludingPrivate::Valid });
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.cpp]: This wasn't in Color::decode

File: Source/WebCore/platform/graphics/Color.cpp [hunk 12]
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
...
+
+        Color::OutOfLineColorData oolcd = {
+            .colorSpace = colorSpace(),
+            .c1 = c1,
+            .c2 = c2,
+            .c3 = c3,
+            .alpha = alpha
+        };
+
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.cpp]: nit: it would be more consistent if this were on the next line.

File: Source/WebCore/platform/graphics/Color.cpp [hunk 13]
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
...
+            .usesFunctionSerialization = flags().contains(FlagsIncludingPrivate::UseColorFunctionSerialization),
+            .data = oolcd
+        };
+
+    } else {
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.cpp]: ditto

File: Source/WebCore/platform/graphics/Color.h [hunk 14]
```diff
@@ -61,10 +61,25 @@ class Color {
         UseColorFunctionSerialization   = 1 << 1,
     };
 
+    struct OutOfLineColorData {
```
Comment (achristensen07) [inline Source/WebCore/platform/graphics/Color.h]: This can be nested in ColorDataForIPC.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

**Review by achristensen07 — APPROVED** [review]
